src/MLMCManager.cpp

```diff
@@ -0,0 +1,450 @@
+        std::cout << "    coarse cost: " << current_cost << std::endl;
+        std::cout << "    uMLMC estimate: " << GetEstimate() << std::endl;
+    }
+
+    if (sample_count_[level] < dump_number_)
+    {
+        std::stringstream ss1;
+        ss1 << "c_upscaled" << sample_count_[level] << ".vector";
+        std::ofstream out1(ss1.str().c_str());
+        InterpolateToFine(fvupscale_, level, sol_coarse).GetBlock(pressure_block).Print(out1, 1);
+    }
+    sample_count_[level]++;
+}
+
+void MLMCManager::CorrectionSample(int level,
+                                   bool verbose)
+{
+    MFEM_ASSERT(level + 1 < num_levels_,
+                "Asking for correction on level that doesn't exist!");
+
+    const int fine_level = level;
+    const int coarse_level = level + 1;
+    const int pressure_block = 1;
+
+    sampler_.NewSample();
+    auto fine_coefficient = sampler_.GetCoefficient(fine_level);
+    fvupscale_.RescaleCoefficient(fine_level, fine_coefficient);
+    auto coarse_coefficient = sampler_.GetCoefficient(coarse_level);
+    fvupscale_.RescaleCoefficient(coarse_level, coarse_coefficient);
+
+    mfem::BlockVector rhs_coarse = RestrictToLevel(fvupscale_, coarse_level, rhs_fine_);
+    mfem::BlockVector sol_coarse(rhs_coarse);
+    fvupscale_.SolveAtLevel(coarse_level, rhs_coarse, sol_coarse);
+    double upscaledq = qoi_.Evaluate(coarse_coefficient, sol_coarse);
+    double temp_cost = fvupscale_.GetSolveTime(coarse_level);
+
+    mfem::BlockVector rhs_fine_local = RestrictToLevel(fvupscale_, fine_level, rhs_fine_);
+    mfem::BlockVector sol_fine(rhs_fine_local);
+    fvupscale_.SolveAtLevel(fine_level, rhs_fine_local, sol_fine);
+    double fineq = qoi_.Evaluate(fine_coefficient, sol_fine);
+    temp_cost += fvupscale_.GetSolveTime(fine_level);
+    // auto error_info = fvupscale_.ComputeErrors(sol_upscaled, sol_fine);
+
+    const double scale = 1.0 / ((double) sample_count_[fine_level] + 1.0);
+    varsum_[fine_level] += scale * ((double) sample_count_[fine_level]) *
+                           ((fineq - upscaledq) - mean_[fine_level]) * ((fineq - upscaledq) - mean_[fine_level]);
+    mean_[fine_level] += scale * ((fineq - upscaledq) - mean_[fine_level]);
+    cost_[fine_level] += scale * (temp_cost - cost_[fine_level]);
+
+    if (verbose)
+    {
+        // ShowErrors(error_info);
+        std::cout << "    finer qoi: " << fineq << ", coarser qoi: " << upscaledq
+                  << ", difference: " << fineq - upscaledq << std::endl;
+        std::cout << "    combined cost: " << temp_cost << std::endl;
+        std::cout << "    uMLMC estimate: " << GetEstimate() << std::endl;
+    }
+
+    if (sample_count_[fine_level] < dump_number_)
+    {
+        // does it make sense to log some information, like the QoI, to go with the visualization?
+        // (conceivably you could even ask qoi_ to make a special picture)
+
+        // for more informative visualization (is it more sensible to do this here or in the python?)
```

Comment:
If you only want log scale for visualization in GLVis, there is an option in GLVis (key "L" if I remember correctly)